File: 102_BinaryTreeLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# 58.31%
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        res = []
        if root is None:
            return res
        queue = []
        queue.append(root)
        while len(queue) != 0:
            queue2 = []
            layer = []
            for tree in queue:
                layer.append(tree.val)
                if tree.left is not None:
                    queue2.append(tree.left)
                if tree.right is not None:
                    queue2.append(tree.right)
            res.append(layer)
            queue = queue2
        return res
	
# Below are three wonderful solutions from LeetCode	
def levelOrder(self, root):
    ans, level = [], [root]
    while root and level:
        ans.append([node.val for node in level])
        LRpair = [(node.left, node.right) for node in level]
        level = [leaf for LR in LRpair for leaf in LR if leaf]
    return ans
	
def levelOrder(self, root):
    ans, level = [], [root]
    while root and level:
        ans.append([node.val for node in level])            
        level = [kid for n in level for kid in (n.left, n.right) if kid]
    return ans
	
def levelOrder(self, root):
    if not root:
        return []
    ans, level = [], [root]
    while level:
        ans.append([node.val for node in level])
        temp = []
        for node in level:
            temp.extend([node.left, node.right])
        level = [leaf for leaf in temp if leaf]
    return ans

File: 102_BinaryTreeLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//3.72%
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int> > res;
		if(!root)
			return res;
		queue<TreeNode *> q;
		queue<TreeNode *> p;
		q.push(root);
		//vector<int> layer;
		while(!q.empty() || !p.empty()){
			vector<int> layer;
			queue<TreeNode *> &qFull = (p.empty()) ? q : p;
			queue<TreeNode *> &pEmpty = (q.empty()) ? q : p;
			while(!qFull.empty()){
				layer.push_back(qFull.front() -> val);
				if(qFull.front() -> left)
					pEmpty.push(qFull.front() -> left);
				if(qFull.front() -> right)
					pEmpty.push(qFull.front() -> right);
				qFull.pop();
			}
			res.push_back(layer);
		}
		return res;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

//Another solution I did before which only need one 'queue'
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) 
	{
        vector<vector<int>> result;
		vector<int> currLayer;
		if(!root)
			return result;
		
		queue<TreeNode*> qNodes;
		TreeNode* currNode;
		qNodes.push(root);
		
		int currNum = 0;
		int nextNum = 1;
		while(!qNodes.empty() && nextNum)
		{
			currNum = nextNum;
			nextNum = 0;
			while(currNum--)
			{
				currNode = qNodes.front();
				currLayer.push_back(currNode -> val);
				qNodes.pop();
				if(currNode -> left)
				{
					nextNum++;
					qNodes.push(currNode -> left);
				}
				if(currNode -> right)
				{
					nextNum++;
					qNodes.push(currNode -> right);
				}
			}
			result.push_back(currLayer);
			currLayer.clear();
		}
		
		return result;
    }
};

//Another solution from LeetCode which use re-cursive
vector<vector<int>> ret;

void buildVector(TreeNode *root, int depth)
{
    if(root == NULL) return;
    if(ret.size() == depth)
        ret.push_back(vector<int>());
    
    ret[depth].push_back(root->val);
    buildVector(root->left, depth + 1);
    buildVector(root->right, depth + 1);
}

vector<vector<int> > levelOrder(TreeNode *root) {
    buildVector(root, 0);
    return ret;
}

// A solution from LeetCode which use a 'NULL' to separate different layers and use only one queue
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        vector<vector<int> >  result;
        if (!root) return result;
        queue<TreeNode*> q;
        q.push(root);
        q.push(NULL);
        vector<int> cur_vec;
        while(!q.empty()) {
            TreeNode* t = q.front();
            q.pop();
            if (t==NULL) {
                result.push_back(cur_vec);
                cur_vec.resize(0);
                if (q.size() > 0) {
                    q.push(NULL);
                }
            } else {
                cur_vec.push_back(t->val);
                if (t->left) q.push(t->left);
                if (t->right) q.push(t->right);
            }
        }
        return result;
    }
};